[mlir][ArmSME] Remove `ConvertIllegalShapeCastOpsToTransposes`

banach-space opened the pull request:
As a follow-up to PR #135841 (see discussion for background), this patch
removes the `ConvertIllegalShapeCastOpsToTransposes` pattern from the SME
legalization pass. This change unblocks folding for ShapeCastOp involving
scalable vectors.

Originally, the `ConvertIllegalShapeCastOpsToTransposes` pattern was introduced
to rewrite certain `vector.shape_cast` ops that could not be lowered otherwise.
Based on local end-to-end testing, this workaround is no longer required, and
the pattern can now be safely removed.

This patch also removes a special case from `ShapeCastOp::fold`, simplifying the
fold logic.

As a side effect of removing `ConvertIllegalShapeCastOpsToTransposes`, we lose
the mechanism that enabled lowering of certain ops like:

```mlir
%res = vector.transfer_read %mem[%a, %b] (...) :
       memref<?x?xf32>, vector<[4]x1xf32>
```
Previously, such cases were handled by:
* Rewriting a nearby `vector.shape_cast` to a `vector.transpose` (via
  `ConvertIllegalShapeCastOpsToTransposes`)
* Then lowering the result with `LiftIllegalVectorTransposeToMemory`.

This patch introduces a new dedicated pattern,
`LowerColumnTransferReadToLoops`, that directly handles illegal
`vector.transfer_read` ops involving leading scalable dimensions.

banach-space committed:
[mlir][ArmSME] Remove `ConvertIllegalShapeCastOpsToTransposes`

As a follow-up to PR #135841 (see discussion for context), this patch
removes `ConvertIllegalShapeCastOpsToTransposes` from the SME legalization
pass and unblocks `ShapeCastOp::fold` for scalable vectors.

AFAIK, `ConvertIllegalShapeCastOpsToTransposes` was originally needed
because we were generating `vector.shape_cast` ops that couldn't be
lowered otherwise. To confirm it's no longer required, I tested this
patch locally using end-to-end tests.

Notably, this also removes a special case from `ShapeCastOp::fold`.

MacDue commented on mlir/test/Dialect/ArmSME/vector-legalization.mlir:
```diff
@@ -491,51 +491,6 @@ func.func @illegal_transpose_no_defining_source_op(%vec: vector<[4]x1xf32>) -> v
-// CHECK-LABEL: @lift_illegal_2d_shape_cast_to_memory
-func.func @lift_illegal_2d_shape_cast_to_memory(%a: index, %b: index, %memref: memref<?x?xf32>) -> vector<1x[4]xf32> {
-  // CHECK: vector.transfer_read {{.*}} : memref<?x?xf32, {{.*}}>, vector<1x[4]xf32>
-  // CHECK-NOT: vector.shape_cast
-  %pad = arith.constant 0.0 : f32
-  %illegalRead = vector.transfer_read %memref[%a, %b], %pad {in_bounds = [false, true]}: memref<?x?xf32>, vector<[4]x1xf32>
-  %cast = vector.shape_cast %illegalRead : vector<[4]x1xf32> to vector<1x[4]xf32>
-  return %cast : vector<1x[4]xf32>
```
I'm not sure what you've tested, but to know if this rewrite is still needed or not this test case should still be possible to lower to LLVM.

banach-space replied:
Thanks Ben!

> I'm not sure what you've tested

I used our e2e tests - from what I can tell, we don't generate such code anymore.

> this test case should still be possible to lower to LLVM

Indeed. @momchil-velikov , since you are working on a generic pattern for "xfer_read with non-trailing scalable dims", could you make sure that this example lowers with your patch?
```mlir
  %illegalRead = vector.transfer_read %memref[%a, %b], %pad {in_bounds = [false, true]}: memref<?x?xf32>, vector<[4]x1xf32>
```

I will wait for Momchil to upload his patch before progressing this one.

momchil-velikov replied:
```
%illegalRead = vector.transfer_read %memref[%a, %b], %pad {in_bounds = [false, true]}: memref<?x?xf32>, vector<[4]x1xf32>
```

I just tried it  and it does not lower. And, AFAICT, it shouldn't, as the last dimension of the memerf (`?`) and the vector (`1`) do not match and the read cannot be inferred to be contiguous, e.g. if we're reading from a memref with dynamic dimensions 4 and 2:

```
[*][]
[*][]
[*][]
[*][]
```

momchil-velikov replied:
This one
```
%illegalRead = vector.transfer_read %memref[%a, %b], %pad {in_bounds = [false, true]}
  : memref<?x2xf32>, vector<[4]x2xf32>
```
is lowered, though (with an implication that `%b` is zero, along the way).

MacDue replied:
> I just tried it and it does not lower. And, AFAICT, it shouldn't, as the last dimension of the memerf

This lowering does not depend on the memref/transpose being contiguous. It lowers the `transfer_read` to a `memref.transpose` + `transfer_read`, which lowers to a loop in the case of a non-contiguous read (such as in this test case): https://godbolt.org/z/b96E7aYq4

banach-space replied:
Sorry, I've not had a chance to return to this yet. It's one of 3 things that are "next" on my list :)

banach-space committed:
fixup! [mlir][ArmSME] Remove `ConvertIllegalShapeCastOpsToTransposes`

Add LowerColumnTransferReadToLoops. Note, this is to address Ben's
comment here:
* https://github.com/llvm/llvm-project/pull/139706/files#r2088605443

banach-space commented:
> I think you could just update the `LiftIllegalVectorTransposeToMemory` rewrite to match `vector.shape_cast` for this case (rather than `vector.transpose`). 

I’d prefer not to, mainly because there’s a fundamental difference between `vector.transpose` and `vector.shape_cast`. The former involves actual data movement, while the latter is not supposed to - as noted in the [docs](https://mlir.llvm.org/docs/Dialects/Vector/#vectorshape_cast-vectorshapecastop):

> It is currently assumed that this operation does not require moving data, and that it will be folded away before lowering vector operations.

This is consistent with recent improvements from @newling - the general goal has been to treat vector.shape_cast as a no-op wherever possible.

> That should lead to less duplication as the existing lowerings for `memref.transpose` and the vector transfer operations would introduce the necessary loops.

Agreed, avoiding duplication is ideal. That said, in this specific case, the duplication introduced is relatively minor, and keeping `shape_cast` logic separate from loop-emitting rewrites aligns better with its semantics.

That said, happy to add TODOs. WDYT?

MacDue reviewed (commented):
> I’d prefer not to, mainly because there’s a fundamental difference between vector.transpose and vector.shape_cast. The former involves actual data movement, while the latter is not supposed to - as noted in the [docs](https://mlir.llvm.org/docs/Dialects/Vector/#vectorshape_cast-vectorshapecastop):

Not blocking this PR (feel free to land it), but I don't really agree with this stance, `vector.shape_cast` has never really been a no-op (the docs have been outdated for a long time). It's no different from any other operation e.g. `vector.transpose`, that may mean data movement, or may fold away (e.g. into a load/store). :slightly_smiling_face:

newling reviewed (approved):
LGTM. I expect my 'canonicalize towards shape_cast' changes will be smoother now, thank you!
